ra/ra_test.go

```diff
@@ -844,8 +848,336 @@ func TestPerformValidationSuccess(t *testing.T) {
+	select {
+	case r := <-va.performValidationRequest:
+		vaRequest = r
+	case <-time.After(time.Second):
+		t.Fatal("Timed out waiting for DummyValidationAuthority.PerformValidation to complete")
+	}
+
+	// Verify that the VA got the request, and it's the same as the others
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Type, vaRequest.Challenge.Type)
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Token, vaRequest.Challenge.Token)
+
+	// Sleep so the RA has a chance to write to the SA
+	time.Sleep(200 * time.Millisecond)
+
+	got, err := ra.SA.GetPausedIdentifiers(ctx, &sapb.RegistrationID{Id: authzPB.RegistrationID}, nil)
+	test.AssertError(t, err, "Should not have any paused identifiers yet, but found some")
+	test.AssertBoxedNil(t, got, "Should have received nil response, but did not")
+	test.AssertMetricWithLabelsEquals(t, ra.pauseCounter, prometheus.Labels{"paused": "false", "repaused": "false", "grace": "false"}, 0)
+
+	// We need the bucket key to scan for in Redis
+	bucketKey, err := ratelimits.NewRegIdDomainBucketKey(ratelimits.FailedAuthorizationsForPausingPerDomainPerAccount, authzPB.RegistrationID, domain)
+	test.AssertNotError(t, err, "Should have been able to construct bucket key, but could not")
+
+	// Verify that a redis entry exists for this identifier:accountID:domain
+	tat, err := redisSrc.Get(ctx, bucketKey)
+	test.AssertNotError(t, err, "Should not have errored, but did")
+
+	// There is no more capacity and the next failed validation will effectively
+	// pause issuance attempts. The ratelimit file is written to increment
+	// capacity every 24 hours, so we can check that the TAT states that, not
+	// that it particularly matters in this context.
+	test.AssertEquals(t, tat, fc.Now().Add(24*time.Hour))
+
+	// A second failed validation should result in the identifier being paused
+	// due to the strict ratelimit.
+	va.PerformValidationRequestResultReturn = &vapb.ValidationResult{
+		Records: []*corepb.ValidationRecord{
+			{
+				AddressUsed:   []byte("192.168.0.1"),
+				Hostname:      domain,
+				Port:          "8080",
+				Url:           fmt.Sprintf("http://%s/", domain),
+				ResolverAddrs: []string{"rebound"},
+			},
+		},
+		Problems: &corepb.ProblemDetails{
+			Detail: fmt.Sprintf("CAA invalid for %s", domain),
+		},
+	}
+
+	challIdx = dnsChallIdx(t, authzPB.Challenges)
+	authzPB, err = ra.PerformValidation(ctx, &rapb.PerformValidationRequest{
+		Authz:          authzPB,
+		ChallengeIndex: challIdx,
+	})
+	test.AssertNotError(t, err, "PerformValidation failed")
+
+	select {
+	case r := <-va.performValidationRequest:
+		vaRequest = r
+	case <-time.After(time.Second):
+		t.Fatal("Timed out waiting for DummyValidationAuthority.PerformValidation to complete")
+	}
+
+	// Verify that the VA got the request, and it's the same as the others.
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Type, vaRequest.Challenge.Type)
+	test.AssertEquals(t, authzPB.Challenges[challIdx].Token, vaRequest.Challenge.Token)
```

Comment:
I don't think we need to confirm this in the test.

Reply:
See below.